shared/src/pool_cache.rs

```diff
@@ -0,0 +1,461 @@
+// - Automatically updating the cache is decoupled from normal pool fetches.
+// A result of this is that it is possible that the same uncached pair is requested multiple times
+// simultaneously and some work is wasted. This is unlikely to happen in practice and the pool is
+// going to be cached the next time it is needed.
+// When pools are requested we mark all those pools as recently used which potentially evicts other
+// pairs from the pair lru cache. Cache misses are fetched and inserted into the cache.
+// Then when the automatic update runs the next time, we request and cache all recently used pairs.
+// For the order book we only care about the "recent" state of the pools. So we can return any
+// result from the cache even if it comes from previous blocks.
+// On the other hand for the driver we need to get the pool at exact blocks which is why we keep a
+// cache of previous blocks in the first place as we could simplify this module if it was only used
+// by the order book.
+
+impl PoolCache {
+    /// number_of_blocks_to_cache: Previous blocks stay cached until the block is this much older
+    /// than the current block.
+    /// number_of_pairs_to_auto_update: The number of most recently used pools to keep track of and
+    /// auto update when the current block changes.
+    pub fn new(
+        number_of_blocks_to_cache: NonZeroU64,
+        number_of_pairs_to_auto_update: usize,
+        inner: Box<dyn PoolFetching>,
+        block_stream: CurrentBlockStream,
+    ) -> Result<Self> {
+        let block = current_block::block_number(&block_stream.borrow())?;
+        Ok(Self {
+            mutexed: Mutex::new(Mutexed::new(number_of_pairs_to_auto_update, block)),
+            number_of_blocks_to_cache,
+            inner,
+            block_stream,
+        })
+    }
+
+    pub async fn update_cache(&self, new_block: u64) -> Result<()> {
+        let pairs = self
+            .mutexed
+            .lock()
+            .unwrap()
+            .recently_used_pairs()
+            .collect::<HashSet<_>>();
+        tracing::debug!("automatically updating {} pair pools", pairs.len());
+        let pools = self
+            .inner
+            .fetch(pairs.clone(), BlockNumber::Number(new_block.into()))
+            .await?;
+        {
+            let mut mutexed = self.mutexed.lock().unwrap();
+            mutexed.insert(new_block, pairs.into_iter(), &pools);
+            let oldest_to_keep = new_block.saturating_sub(self.number_of_blocks_to_cache.get() - 1);
+            mutexed.remove_cached_blocks_older_than(oldest_to_keep);
+            mutexed.last_update_block = new_block;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl PoolFetching for PoolCache {
+    async fn fetch(&self, pairs: HashSet<TokenPair>, block: BlockNumber) -> Result<Vec<Pool>> {
+        // We assume that if "Latest" is used it is okay to return an earlier cached version even if
```

Comment:
I assume you don't want to pass `Option<BlockNumber>` because this would make the PR too large? Can we do this right after (could have also been a preparational PR before)?

Reply:
Yes I didn't want to include it in this PR. Good idea to do it first. Will make a PR for that now.